driver-core/src/test/functional/com/mongodb/test/extension/FlakyTestExtension.java

```diff
@@ -0,0 +1,178 @@
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb.test.extension;
+
+
+import com.mongodb.test.FlakyTest;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.AfterTestExecutionCallback;
+import org.junit.jupiter.api.extension.ConditionEvaluationResult;
+import org.junit.jupiter.api.extension.ExecutionCondition;
+import org.junit.jupiter.api.extension.Extension;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.TestInstantiationException;
+import org.junit.jupiter.api.extension.TestTemplateInvocationContext;
+import org.junit.jupiter.api.extension.TestTemplateInvocationContextProvider;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.platform.commons.util.Preconditions;
+
+import java.lang.reflect.Method;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Spliterator;
+import java.util.stream.Stream;
+
+import static com.mongodb.test.FlakyTest.CURRENT_REPETITION_PLACEHOLDER;
+import static com.mongodb.test.FlakyTest.DISPLAY_NAME_PLACEHOLDER;
+import static com.mongodb.test.FlakyTest.TOTAL_REPETITIONS_PLACEHOLDER;
+import static java.util.Collections.singletonList;
+import static java.util.Spliterators.spliteratorUnknownSize;
+import static java.util.stream.StreamSupport.stream;
+import static org.junit.platform.commons.util.AnnotationUtils.findAnnotation;
+import static org.junit.platform.commons.util.AnnotationUtils.isAnnotated;
+
+
+/**
+ * A {@code TestTemplateInvocationContextProvider} that supports the {@link FlakyTest @FlakyTest} annotation.
+ */
+public class FlakyTestExtension implements TestTemplateInvocationContextProvider, AfterTestExecutionCallback {
+
+    private int maxAttempts = 0;
+    private FlakyTestDisplayFormatter formatter;
+    private Boolean testHasPassed;
+
+
+    @Override
+    public void afterTestExecution(final ExtensionContext extensionContext) {
+        testHasPassed = extensionContext.getExecutionException().map(e -> e instanceof TestInstantiationException).orElse(true);
+    }
+
+    @Override
+    public boolean supportsTestTemplate(final ExtensionContext context) {
+        return isAnnotated(context.getTestMethod(), FlakyTest.class);
+    }
+
+    @Override
+    public Stream<TestTemplateInvocationContext> provideTestTemplateInvocationContexts(final ExtensionContext context) {
+        Method testMethod = context.getRequiredTestMethod();
+        String displayName = context.getDisplayName();
+
+        if (isAnnotated(testMethod, Test.class)) {
```

Comment:
This just helps ensures that tests aren't annotated with multiple test templates. It doesn't catch all possibilities but these are the two most common.